sw/device/tests/rstmgr_sw_rst_ctrl_test.c

```diff
@@ -0,0 +1,327 @@
+#include "sw/device/lib/dif/dif_spi_device.h"
+#include "sw/device/lib/dif/dif_spi_host.h"
+#include "sw/device/lib/dif/dif_usbdev.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_framework/ottf.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "i2c_regs.h"
+#include "spi_device_regs.h"
+#include "spi_host_regs.h"
+#include "usbdev_regs.h"
+
+/**
+ *  RSTMGR SW_RST_CTRL Test
+ *
+ *  This test checks RSTMGR.SW_RST_CTRL_N[index] with peripheral devices.
+ *  There are 8 RSTMGR.SW_RST_CTRL_N registers and each register
+ *  controls peripheral reset as follows:
+ *
+ * // index | device     |  test register |  reset value |  prgm value |
+ * // ------------------------------------------------------------------
+ * // 0     | SPI_DEVICE |  CFG           |  0x7f00      |  0x3f0c
+ * // 1     | SPI_HOST0  |  CONFIGOPTS    |  0x0         |  0x3210000
+ * // 2     | SPI_HOST1  |  CONFIGOPTS    |  0x0         |  0x6540000
+ * // 3     | USB        |  EP_OUT_ENABLE |  0x0         |  0xc3
+ * // 4     | USBIF      |  RXENABLE_OUT  |  0x0         |  0x695
+ * // 5     | I2C0       |  TIMING0       |  0x0         |  0x8b00cfe
+ * // 6     | I2C1       |  TIMING1       |  0x0         |  0x114010d8
+ * // 7     | I2C2       |  TIMING2       |  0x0         |  0x19ec1595
+ *
+ * 'test register' is a rw type register under each peripheral device.
+ * During the test, these registers are programmed with arbitrary values. ('prgm
+ * value') when sw resets are asserted, those values are updated to reset value.
+ * In each round, the rstmgr asserts reset to each peripheral device and
+ * each test register value is compared with the expected reset value.
+ */
+
+#define kRstMgr kTopEarlgreyResetManagerSwResetsLast + 1
+#define kNumNodes kRstMgr + 1
+
+typedef struct node node_t;
+typedef void config_f(node_t *);
+
+// peripheral device template
+// name : device name
+// base : base address
+// offset : offset to the test register
+// handler : dif handler for each peripheral device
+// config_f : function ptr to each peripheral program
+// reset_val : reset value for comparison
+// prgm_val : expected value for test register after configuration
+//            function is called.
+struct node {
+  const char *name;
+  uintptr_t base;
+  ptrdiff_t offset;
+  void * handler;
+  config_f *config;
+  uint32_t reset_val;
+  uint32_t prgm_val;
+};
+
+// handler declaration
+dif_rstmgr_t rstmgr;
+dif_spi_device_t spi_dev;
+dif_spi_host_t spi_host0, spi_host1;
+dif_usbdev_t usbdev;
+dif_i2c_t i2c0, i2c1, i2c2;
+
+// pre declaration of template
+static node_t test_node[kNumNodes];
+
+// test register read from a given peripheral device
+uint32_t read_test_reg(node_t *node) {
+  return (mmio_region_read32(mmio_region_from_addr(node->base), node->offset));
+}
+
+// init function
+void init_nodes(int n) {
```

Comment:
I will recommend not passing this argument. See comment below.

Reply:
removed